### ISSUE TITLE
Add an 'is-enabled' state for Units with Also= states different from self

### ISSUE THREAD

**Issue by skylerbunny** [systemd/systemd#6681]
### Submission type

 - Bug report

### Used distribution

Centos 7.3

### In case of bug report: Expected behaviour you didn't see

Unit is disabled but an Also= Unit is enabled:
```systemctl is-enabled``` returns 'inconsistent'. (or something)
Unit is enabled but an Also= Unit is disabled:
```systemctl is-enabled``` returns 'inconsistent'. (or something)

### In case of bug report: Unexpected behaviour you saw

Unit is disabled but an Also= Unit is enabled:
```systemctl is-enabled``` returns 'disabled'.
Unit is enabled but an Also= Unit is disabled:
```systemctl is-enabled``` returns 'enabled'.

### In case of bug report: Steps to reproduce the problem

1. Install Centos 7.3. Install rpcbind service (which has an 'indirect' state out of the box for rpcbind.service).
2. Add any WantedBy= or RequiredBy= dependency you like to the rpcbind.service Install section of the unit file, and issue a ```systemctl daemon-reload```.
3. Behavior described above occurs if rpcbind.socket's enable/disable state is modified, since it doesn't have an Also= section in its own Unit file.

Issuing a ```systemctl enable``` to a Unit file that has both a WantedBy and Also section will enable both the unit file itself, and the Also unit. Conversely, issuing a disable will do the above, disabling both. That's great. The problem is, if a unit file which refers to another via Also is queried with ```is-enabled```, and its state is different than the other(s), it gives no indication of this fact. **It should, because from an automation point of view, this is a state that requires action; all services in an Also=section should have the same state as the calling unit if the calling Unit has a WantedBy/RequiredBy.** This is why I'm terming this as a bug rather than an RFE.

See
https://tickets.puppetlabs.com/browse/PUP-7163 (Puppet)
https://github.com/ansible/ansible/pull/28274 (Ansible)

The tl;dr of this is that there doesn't appear to be a clean command line method of querying a unit to determine:
a. That it has an Also= dependency, if it also has a WantedBy or RequiredBy -- technically this isn't the real issue, but it's hidden, making it impossible to deduce (b) in automation -- more importantly:
b. Whether or not Units referred to in Also= have states different from that Unit file, which may be of interest/concern to a user (and automation scripts want to be able to rely on/query). 'indirect' tells us (a), but ONLY if the Unit is disabled and has no references other than Also=. As explained above, this isn't really the whole story.

**Comment by skylerbunny** [comment]
I was pointed to https://bugzilla.redhat.com/show_bug.cgi?id=864298 which gives me some historical context for the bug I've filed. Now that I know 'where systemd came from' in this regard, I'd like to make a specific recommendation for a fix.

Originally, according to the bug above, Unit files with only an ```Also=``` (but no ```WantedBy=``` or ```RequiredBy=```) simply came back with a status of ```static```, even though when one runs ```systemctl (dis|en)able```, actions are taken on the ```Also=``` Units' symlinks. So ```indirect``` was created to provide a distinction for these services.

IMHO, while this helped to some extent, it doesn't go far enough, given what ```systemctl (dis|en)able``` **actually does** when run. This is the logic matrix of ```systemctl is-enabled```  that I propose:

--
* Retire the 'indirect' status entirely. Then...
* If a Unit has *only* an ```Also=```, we know that executing ```systemctl enable``` or ```systemctl disable``` will act on any ```Also=``` unit. Therefore...
  * Check the enabled status of all ```Also=``` units, run recursively\*\*. The status returned for ```is-enabled``` should be 'enabled' if all of them are enabled, 'disabled' if all of them are disabled, and 'split'* if some are enabled or disabled.
* If a Unit has an ```Also=``` *and* (```WantedBy=``` or ```RequiredBy=```)
  * Check the enabled status of the Unit itself, and add this to the list of statuses for all ```Also=``` units, run recursively\*\*. Then, as above, the status returned for ```is-enabled``` should be 'enabled' if all of them are enabled, 'disabled' if all of them are disabled, and 'split'* if some are enabled or disabled.

\* I don't really care what the status name is as much as that it should be *distinct from enabled or disabled, and that 'indirect' makes no sense*. It could be 'split', 'inconsistent', 'both', 'multiple'...whatever.

\*\* This is exactly what ```systemctl (dis|en)able``` does when run against a Unit anyway to add or remove symlinks. It follows that ```is-enabled``` should mirror this behavior.

--

Since 'split' represents a unit or units that has some of its services enabled, ```systemctl is-enabled``` for these services would return ```0```, as it does for ```enabled``` and ```indirect```. Generally speaking, this would mean that the return code is still consistent with what it does today, and (again as it is today anyway) one still has to use the stdout of ```is-enabled``` to dig deeper.

Doing this means automation processes like Ansible, Chef and Puppet (or a user) can be certain of the state of such units based on ```systemctl is-enabled```. One can then determine conclusively whether an enable or disable action must subsequently be run, and in hindsight whether a change has been made.

**Comment by hlovdal** [comment]
Any plans for fixing this? Being able to automate installs is kind of important.

**Comment by cjsoftuk** [comment]
I've just followed down a rabbit hole and found this troubleshooting Puppet always changing a service on a puppet run (notably lightdm on Buster, which installs as display-manager.service) and it seems to lead all the way back to this...... I'd really love to help fix this, but I don't even know where to start.
